### PR TITLE
15 django 2.0

### PR DESCRIPTION
**Pull request by frankwiles**
There are a lot of small Django 2.x related changes in here, but the largest change is the removal of Thorn and replacing it with our own webhook dispatcher. 

This ended up being the _simplest_ approach and gives us some added ops related benefits (since we are tracking success/failure a bit on the API end) which should help avoid situations like: 

- subscribing localhost
- subscribing non-SSL endpoints
- being able to detect webhooks that are failing from the Django admin / API and not just from the model service end of things 

Happy for any feedback anyone wants to provide.

### CODE REVIEW

**Comment by janeeisenstein** [comment]
@frankwiles Can resolve the conflicts this branch has or should I just test against the branch for now? Also, will your subscribing changes force developers to run the client apps in docker? If so, lets leave them for a second stage revision after all our current apps have been upgrade to only (?) run in docker?

**Comment by frankwiles** [comment]
@janeeisenstein conflicts resolved 

And no, nothing about these changes _forces_ Docker, but we will be moving everything to _assume_ docker as part of these docs, updates, upgrades, etc.

**Comment by janeeisenstein** [comment]
@frankwiles I agree we're moving to a docker only setup. However, my current webhooks_subscriber table is full of 'http://localhost:8080/callback' entries.

**Comment by frankwiles** [comment]
Ah, I don't think that will be be an issue for already subscribed callbacks.  If you run into problems with registration just set: 

```python
SIMPL_WEBHOOKS_ALLOW_LOCALHOST = True 
SIMPL_WEBHOOKS_ALLOW_HTTP = True
```

in your settings. They just are off by default.

**Comment by janeeisenstein** [comment]
@frankwiles I'm getting this error running ./manage.py migrate against my existing Simpl DB
	CommandError: Conflicting migrations detected; multiple leaf nodes in the migration graph: (0042_auto_20191216_1732, 0044_auto_20200418_2224 in games).
	To fix them run 'python manage.py makemigrations --merge'
	(simpl-games-api-upgrade) softweave-2020:simpl-games-api-upgrade janee$ python manage.py makemigrations --merge
	Merging games
	  Branch 0042_auto_20191216_1732
	    - Alter unique_together for decision (1 constraint(s))
	  Branch 0044_auto_20200418_2224
	    - Alter unique_together for decision (1 constraint(s))
	    - Alter field data on game
	    - Alter field data on period
	    - Alter field data on result
	    - Alter field data on role
	    - Alter field data on run
	    - Alter field data on runuser
	    - Alter field data on scenario
	    - Alter field data on world
	    - Alter field data on decision

	Merging will only work if the operations printed above do not conflict
	with each other (working on different fields or models)
	Do you want to merge these migration branches? [y/N]

**Comment by janeeisenstein** [comment]
@frankwiles I'm getting this error running ./manage.py runserver 0.0.0.0:8100

	2020-05-27 12:58:41,041 INFO autoreload: Watching for file changes with StatReloader
	Performing system checks...

	System check identified no issues (0 silenced).
	May 27, 2020 - 12:58:41
	Django version 2.2.12, using settings 'config.settings.local'
	Starting development server at http://0.0.0.0:8100/
	Quit the server with CONTROL-C.
	2020-05-27 12:59:51,217 ERROR log: Internal Server Error: /apis/runs/
	Traceback (most recent call last):
	  File "/Users/janee/.virtualenvs/simpl-games-api-upgrade/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
	    return self.cursor.execute(sql, params)
	psycopg2.errors.UndefinedColumn: column webhooks_subscriber.connected does not exist
	LINE 1: ...bscriber"."url", "webhooks_subscriber"."user_id", "webhooks_...
	                                                             ^

**Comment by janeeisenstein** [comment]
The runserver error occurred after I merged and applied the migrations.

**Comment by janeeisenstein** [comment]
@frankwiles Solved last problem by rerunning makemigrations and migrate. Will continue testing.

**Comment by janeeisenstein** [comment]
@frankwiles Actually the error is still there and is activated by running a modelservice management command against the api. For example, running simpl-calc-model create_default_env causes this error on the api server:

	2020-05-27 14:10:56,063 ERROR log: Internal Server Error: /apis/games/simpl-calc/
	Traceback (most recent call last):
	  File "/Users/janee/.virtualenvs/simpl-games-api-upgrade/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
	    return self.cursor.execute(sql, params)
	psycopg2.errors.UndefinedColumn: column webhooks_subscriber.connected does not exist
	LINE 1: ...bscriber"."url", "webhooks_subscriber"."user_id", "webhooks_...

**Comment by janeeisenstein** [comment]
@fcurella This is not ready for merging to master. Too many problems remain.

**Comment by janeeisenstein** [comment]
@frankwiles Please don't merge to master until this branch is capable of updating existing Simpl databases.

**Comment by fcurella** [comment]
@janeeisenstein @frankwiles I've converted the PR to a Draft for the time being. This will prevent accidental merges.

File: simpl/games/apps.py
Comment (jefftriplett) [inline simpl/games/apps.py:4]: Style: I'd remove all of the commented out lines once you are sure it does what you want.

Comment (frankwiles) [reply simpl/games/apps.py]: Ooh good catch, I found a couple other thorn comments to remove as well.

**Comment by frankwiles** [comment]
@janeeisenstein yeah I wasn't running migrations against an existing pre-change database.  I'll look into that next and fix.

**Comment by janeeisenstein** [comment]
@frankwiles The README needs to be updated starting with the Django version requirements. How to run locally using Docker needs to be added. Running locally outside Docker needs to document changing these variables in config/settings/local:
    SIMPL_WEBHOOKS_ALLOW_LOCALHOST = True 
    SIMPL_WEBHOOKS_ALLOW_HTTP = True

File: config/settings/common.py
Comment (janeeisenstein) [inline config/settings/common.py:51]: Why is rest_auth a LOCAL_APP?

**Comment by janeeisenstein** [comment]
@frankwiles Once simpl-games-api no longer depends on thorn, we can remove thorn from simpl-modelservice requirements (see https://github.com/simplworld/simpl-modelservice/issues/29). Can this be done independently of simpl-games-api webhooks or should there be a shared simpl-webhooks package used by both projects?

File: requirements.txt
Comment (janeeisenstein) [inline requirements.txt:3]: How about locking down only to minor version number (django~=2.2.0) so we won't miss security patches?
The commented out old requirement is no longer needed.

File: requirements.txt
Comment (janeeisenstein) [inline requirements.txt:12]: Needs to be updated to simpl-users v0.8.2, to support TokenAuthentication. Also, the 0.8 indicates the Simpl package supports Django 2.2.

**Comment by jefftriplett** [comment]
I approved, based on code I see not necessarily because it's ready to be merged given previous feedback that y'all are working for.

**Comment by janeeisenstein** [comment]
@frankwiles with the renaming of the webhook directory, these settings also need to be renamed.
SIMPL_WEBHOOKS_ALLOW_LOCALHOST -> SIMPL_WEBHOOK_ALLOW_LOCALHOST
SIMPL_WEBHOOKS_ALLOW_HTTP  -> SIMPL_WEBHOOK_ALLOW_HTTP

**Comment by janeeisenstein** [comment]
@frankwiles I dropped and recreated my local Simpl database and initialized it for simpl-calc-model and pop-model using the current simpl-games-api master branch code. After switching to using your latest changes and adding SIMPL_WEBHOOKS_ALLOW_LOCALHOST = True and  SIMPL_WEBHOOKS_ALLOW_HTTP = True to settings/local.py, I get this error attempting to start simpl-calc-model:
	WARNING log: Not Found: /apis/hooks/
	WARNING basehttp: "POST /apis/hooks/ HTTP/1.1" 404 13074

**Comment by frankwiles** [comment]
Yep, I realized this morning I'm missing then subscription API entirely in the setup.  Working on that now. I'd hold off on trying to review this WIP until I think it's ready.

**Comment by janeeisenstein** [comment]
@frankwiles  I'll hold off further testing until you give notice. BTW the migrate scripts tested out fine.

**Comment by frankwiles** [comment]
@janeeisenstein awesome.  Yeah the issue ended up being Thorn ends up being `webhooks.subscriber` in the migrations table and so was `simpl.webhooks.subscriber` so I just renamed it it `webhook` (singular) to get around that for now.  

I didn't spot it originally as I was starting with empty databases.

**Comment by janeeisenstein** [comment]
@frankwiles Let me know when this PR is ready for testing.

**Comment by frankwiles** [comment]
Ok @janeeisenstein and @fcurella this is ready for review if either of you have a moment!

File: simpl/games/events.py
Comment (fcurella) [inline simpl/games/events.py:9]: Contrary to this comment, I'm not seeing the User model in the set

File: simpl/games/tests/test_api_views.py
Comment (fcurella) [inline simpl/games/tests/test_api_views.py:334]: remember to remove this or convert to proper logging

File: simpl/webhook/tasks.py
Comment (fcurella) [inline simpl/webhook/tasks.py:25]: Do we also want to add retries for just network issues?

File: start.sh
Comment (fcurella) [inline start.sh:4]: remember to delete this

**Comment by janeeisenstein** [comment]
@frankwiles Things don't look good running simpl-games-api with SIMPL_WEBHOOKS_ALLOW_LOCALHOST = True
SIMPL_WEBHOOKS_ALLOW_HTTP = True. The simpl-calc-model 's **create_default_env** doesn't run through to completion. It fizzles out trying to create the run.
Please advise.

**Comment by janeeisenstein** [comment]
@frankwiles There is definitely something funky with the current state of this branch. I dropped and recreated my local Simpl database using the master branch and was able to run simpl-calc-model 's create_default_env against it with no problems.

**Comment by frankwiles** [comment]
Hmmm ok I was able to run it with one of the sim, I think it might have been Blackjack or ROE, but I'll give it another whirl to verify things.

**Comment by janeeisenstein** [comment]
@frankwiles I retested simpl-calc-model 's create_default_env against this branch and it had no problems.

**Comment by janeeisenstein** [comment]
@frankwiles Running locally, REST requests sometimes do not reach the simpl-games-api server. The attached file describes the steps I followed today that demonstrate this problem. Earlier, I've seen this problem when trying to login and when running create_default_env. These problems do not occur when using the simpl-games-api master branch.

[20200730-test.txt](https://github.com/simplworld/simpl-games-api/files/5001890/20200730-test.txt)

File: simpl/games/events.py
Comment (frankwiles) [reply simpl/games/events.py]: Ah it's actually handled specifically below, but documented it just now for clarity.

File: simpl/webhook/tasks.py
Comment (frankwiles) [reply simpl/webhook/tasks.py:26]: I think we should maybe consider that in the future, I'm a little worried about "out of order" update scenarios and how that might impact the modelservice's "view" of what the up to date data is.

**Comment by frankwiles** [comment]
@janeeisenstein @fcurella ok I found yet another Celery related bug (and a few other small things) and have this working with simpl-calc again locally.  

I think it's ready to go.

**Comment by janeeisenstein** [comment]
@frankwiles Does celery have to be running to use this branch? If so, please update README with steps to run this branch locally.

**Comment by janeeisenstein** [comment]
@frankwiles Congratulations! I verified I can run pop locally. There is something off with simpl-user migrations. Details are in the attached file. I'll next test pop and simpl-admin against this branch.

[20200817-test-1.txt](https://github.com/simplworld/simpl-games-api/files/5086228/20200817-test-1.txt)

**Comment by janeeisenstein** [comment]
@frankwiles Ran into problems testing pop against this branch. Details are in the attached file.  I'll next recreate the simpl database using this branch and retest pop.

[20200817-test-2.txt](https://github.com/simplworld/simpl-games-api/files/5086373/20200817-test-2.txt)

**Comment by janeeisenstein** [comment]
@frankwiles After recreating the simpl database, still cannot advance from Setup to Play phase. Will retest against a simpl database created/accessed using the master branch.

**Comment by frankwiles** [comment]
Closer it seems! Thanks for diving in and checking things out.  I'll recreate these and fix.

**Comment by janeeisenstein** [comment]
@frankwiles FWIW I also had a subjective impression that simpl-calc's receiving of database updates was slower with this branch. After retesting with this master branch today, I believe it would be worth instrumenting simpl-calc-ui to time how long it takes to receive database updates using both branches. Delaying change notifications is something we of course want to avoid.

**Comment by janeeisenstein** [comment]
@frankwiles 
1. Can we run `simpl-games-api` in Docker without also running the apps that communicate with it in Docker? 
2. Why are Docker users instructed to set up a superuser and a staff user? Previously only the default simpl@simpl.world super user was needed for local development.

**Comment by janeeisenstein** [comment]
@frankwiles If this branch requires Python 3.7 as indicated in the README, please update the .travis.yml.

**Comment by janeeisenstein** [comment]
@frankwiles FWIW running make shell results in 
```
Makefile:3: .version: No such file or directory
make: *** No rule to make target `.version'.  Stop.
```

**Comment by frankwiles** [comment]
> @frankwiles
> 
> 1. Can we run `simpl-games-api` in Docker without also running the apps that communicate with it in Docker?

Yes.  Compose is setup to expose the API on localhost:8100 so you'd just need to adjust your model and UIs to point there. 

> 2. Why are Docker users instructed to set up a superuser and a staff user? Previously only the default [simpl@simpl.world](mailto:simpl@simpl.world) super user was needed for local development.

Mostly because it's not really a good practice to have YOUR (as in you the human) user account be the same as a service/machine account (the simpl auth account).  I tend to setup `frank@revsys.com` as my superuser locally always and then we just need a staff user for the model bits.  Also avoids any possible future mistakes of having `simpl@simpl.world` be setup as a superuser and either introducing authorization mistakes (something should be staff only but is superuser only or vice versa) or someone compromising at a superuser level the API in a deployed environment because the creds were published in a repo.

**Comment by janeeisenstein** [comment]
@frankwiles I was able to run `make shell` successfully. Next, I ran `./manage.py migrate` and  `./manage.py create_simpl_user` (to avoid having to override with our Simpl apps' default local settings).
Running pop-model `./manage.py create_default_env` resulted in:
```
Traceback (most recent call last):
  File "/Users/janee/.virtualenvs/pop-model/lib/python3.6/site-packages/urllib3/connection.py", line 157, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw
  File "/Users/janee/.virtualenvs/pop-model/lib/python3.6/site-packages/urllib3/util/connection.py", line 84, in create_connection
    raise err
  File "/Users/janee/.virtualenvs/pop-model/lib/python3.6/site-packages/urllib3/util/connection.py", line 74, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 61] Connection refused
```
Possibly due to needing to override settings for `SIMPL_WEBHOOK_ALLOW_LOCALHOST` and `SIMPL_WEBHOOK_ALLOW_HTTP`?

**Comment by frankwiles** [comment]
@janeeisenstein bah yeah I completely forgot about the webhooks when I told you could run it in docker and not the model service, Docker doesn't provide a super easy way to go from inside Docker to an outside service like that.  Sorry I wasted your time, but yeah the model and UI are best done in docker as well. 

I took pop-model for a spin based on your issues above, in docker (I committed the docker-compose for it), and I couldn't reproduce them.

**Comment by janeeisenstein** [comment]
> @janeeisenstein bah yeah I completely forgot about the webhooks when I told you could run it in docker and not the model service, Docker doesn't provide a super easy way to go from inside Docker to an outside service like that. Sorry I wasted your time, but yeah the model and UI are best done in docker as well.
> 
> I took pop-model for a spin based on your issues above, in docker (I committed the docker-compose for it), and I couldn't reproduce them.

@frankwiles I usually try out my README documentation before committing.  It's sounding like we will not be able to acceptance test this branch unless all client apps are running in docker.

**Comment by janeeisenstein** [comment]
@frankwiles If you only ran pop-model in docker, you did not test this issue I reported 8 days ago: "cannot login as leader@pop.edu and advance a run from Setup to Play phase."

**Comment by frankwiles** [comment]
Ok I _**FINALLY**_ found the source of my issue.  Turns out, strangely enough, the string `updated` isn't the same as `changed` grrrr can't believe it took me so long to find that.

**Comment by janeeisenstein** [comment]
Verified now able to change run phase from POP leader dashboard and from Simpl Admin. Bravo!

**Review by janeeisenstein — APPROVED** [review]

**Comment by frankwiles** [comment]
Whoohooo!

**Pull request merged**
